### ISSUE TITLE
Geometries without ExtraTower

### ISSUE THREAD

**Issue by spradlin** [spradlin/WCSim#31]
In the [18 May 2023 HK software meeting](https://hkdbweb.in2p3.fr/meetingplanner.htm?meetingid=A00009000), i asked about the possibility of defining the HyperK geometries in such a way that the ExtraTower is not constructed.  Alors, what would it take to do that?

**Comment by spradlin** [comment]
## Condition for ExtraTower construction

The [condition for constructing an ExtraTower](https://github.com/tdealtry/WCSim/blob/fc243cb6dea0bd51ec2c9758a9c48829f58f161e/src/WCSimConstructCylinder.cc#L546) is
```
if(!(WCBarrelRingNPhi*WCPMTperCellHorizontal == WCBarrelNumPMTHorizontal))
```

`WCBarrelRingNPhi` is [computed from the other two terms](https://github.com/tdealtry/WCSim/blob/fc243cb6dea0bd51ec2c9758a9c48829f58f161e/src/WCSimConstructCylinder.cc#L82-L83) by integer division
```
WCBarrelRingNPhi     = WCBarrelNumPMTHorizontal/WCPMTperCellHorizontal; 
```

`WCPMTperCellHorizontal` and `WCBarrelNumPMTHorizontal` are defined in the [`WCSimDetectorConfigs.cc`](https://github.com/tdealtry/WCSim/blob/fc243cb6dea0bd51ec2c9758a9c48829f58f161e/src/WCSimDetectorConfigs.cc) function that defines the geometry.  They can be either simple numbers or derived from other quantities.  For the [`HyperK_HybridmPMT_WithOD` geometry](https://github.com/tdealtry/WCSim/blob/fc243cb6dea0bd51ec2c9758a9c48829f58f161e/src/WCSimDetectorConfigs.cc#L755-L878) they are derived:
```
  WCSimPMTObject * PMT = CreatePMTObject("BoxandLine20inchHQE", WCIDCollectionName);
  :
  WCPMTRadius         = PMT->GetRadius();
  :
  mPMT_vessel_radius = 254.*CLHEP::mm;
  :
  WCIDDiameter          = 64.8*m;
  :
  if(WCPMTPercentCoverage==-1.) WCPMTPercentCoverage  = 20.0;
  if(WCPMTPercentCoverage2==-1.) WCPMTPercentCoverage2  = 5.0;
  WCPMTperCellHorizontal = std::lround((WCPMTPercentCoverage+WCPMTPercentCoverage2) / (WCPMTPercentCoverage2==0?(WCPMTPercentCoverage+WCPMTPercentCoverage2):WCPMTPercentCoverage2)); 
   :
  WCBarrelNumPMTHorizontal = std::lround(WCIDDiameter*sqrt(pi*(WCPMTPercentCoverage+WCPMTPercentCoverage2))/(10.0*std::max(WCPMTRadius,mPMT_vessel_radius)));
```
The conditional setting of `WCPMTPercentCoverage` and `WCPMTPercentCoverage2` seems weird.  And it uses `==` comparison of floating points, which is definitely bad.  These two variables are [initialized to -1. by default](https://github.com/tdealtry/WCSim/blob/a51f2a6b9afb393d8010e88cd4d8d391952f8020/include/WCSimDetectorConstruction.hh#L604-L605), but this should really be revised.

Anyway, with `WCPMTPercentCoverage  = 20.0` and `WCPMTPercentCoverage2  = 5.0`,
```
WCPMTperCellHorizontal = std::lround((WCPMTPercentCoverage+WCPMTPercentCoverage2) / WCPMTPercentCoverage2); 
    = std::lround((20.0 + 5.0) / 5.0)
    = 5
```
which is consistent with what i have seen in visualizations (see some of the pictures on https://github.com/spradlin/WCSim/issues/24).

For `WCBarrelNumPMTHorizontal`, i need the [radius of the `BoxandLine20inchHQE` PMT](https://github.com/tdealtry/WCSim/blob/a51f2a6b9afb393d8010e88cd4d8d391952f8020/src/WCSimPMTObject.cc#L2116):
```
G4double BoxandLine20inchHQE::GetRadius() {return .254*m;}
```

So, running the arithmetic,
```
WCBarrelNumPMTHorizontal = std::lround(WCIDDiameter*sqrt(pi*(WCPMTPercentCoverage+WCPMTPercentCoverage2))/(10.0*std::max(WCPMTRadius,mPMT_vessel_radius)));
    = std::lround(64.8m * sqrt(pi * (20.0 + 5.0)) / (10. * max(.254m, 254mm)) )
    = std::lround(64.8m * 5.0 * sqrt(pi) / 2.54m)
    = std::lround(226.09253846196344)
    = 226
```

Thus the integer arithmetic gives
```
WCBarrelRingNPhi     = WCBarrelNumPMTHorizontal/WCPMTperCellHorizontal; 
    = 226/5
    = 45
```
with a remainder
```
WCBarrelNumPMTHorizontal % WCPMTperCellHorizontal = 226 % 5
    = 1
```

**Comment by spradlin** [comment]
## Avoiding ExtraTower in HyperK_HybridmPMT_WithOD

So if we want to adjust some of the numbers to avoid the ExtraTower condition, what do we have free?  I assume that the following dimensions are fixed by the physical dimensions of the PMTs:
- WCPMTRadius
- mPMT_vessel_radius

I will also assume that the hard-coded factor 10.0 in the denominator of the `WCBarrelNumPMTHorizontal` computation is hard-coded to convert the `%` of the `WCPMTPercentCoverage` to fractions (when the sqrt is considered).

The impact of modifying a bulk geometric parameter, like the `WCIDDiameter` would need to be carefully considered.

This leaves the following factors:
- WCPMTPercentCoverage
- WCPMTPercentCoverage2

If the coverage figures are modified, then it would be most convenient to change them in concert so that `WCPMTperCellHorizontal` remains 5.

### Tweaks to coverage figures

Simultaneously scaling `WCPMTPercentCoverage` and `WCPMTPercentCoverage2` down by
```
(225./226.)**2 = 0.9911700211449604
```
should do it.
Or, more precisely
```
(225./226.09253846196344)**2 = 0.9903588260529245
```
```
WCPMTPercentCoverage  = 20.0 * (225./226.09253846196344)**2 = 19.80717652105849
WCPMTPercentCoverage2  = 5.0 * (225./226.09253846196344)**2 = 4.951794130264623
```

Alternatively, they could be scaled up by `(230./226.)**2`:
```
(230./226.09253846196344)**2 = 1.0348638399644388
WCPMTPercentCoverage  = 20.0 * (230./226.09253846196344)**2 = 20.697276799288776
WCPMTPercentCoverage2  = 5.0 * (230./226.09253846196344)**2 = 5.174319199822194
```

**Comment by tdealtry** [comment]
Thanks for this quick work Patrick! I'd be interested in seeing the nPMTs for the two cases

FYI your confusion about these
```
if(WCPMTPercentCoverage==-1.) WCPMTPercentCoverage  = 20.0;
if(WCPMTPercentCoverage2==-1.) WCPMTPercentCoverage2  = 5.0;
```
The idea whoever coded it wanted here was, if people want to modify the photocoverage, they set the photocoverage parameters first. See [here](https://github.com/tdealtry/WCSim/blob/main_into_hybrid/WCSim.mac#L40)

But of course the floating point equality is a terrible idea as you say (and something I've never spotted). I'll make a fix

**Comment by tdealtry** [comment]
Just to say that it works! As in the extra tower isn't built. Thank you!

As for number of 20" PMTs / mPMT modules
* Original 18952 / 4716
* Rounding down 18864 / 4686
* Rounding up 20296 / 5090

I'm going to fiddle & find something that makes the mPMT coverage closer to reality, and move to that for the default.
I'll also do the same for the SK & IWCD geometries

**Comment by tdealtry** [comment]
I've cleaned up the default for HK FD & SK (IWCD has no extra tower by default). I'm running my standard set of validations - will report the result tomorrow. Fingers crossed for a strong reduction in killed tracks

I've also gone further (too far?) & made constructing a geometry with an extra tower fatal. I've also included a script to calculate a valid photocoverage.

**Comment by tdealtry** [comment]
Coming back to this after too long...

The extra tower isn't all that's causing the killed tracks

# Geometry test 1
Using the following
  - 19942 20" (close to 20% coverage)
  - 1550 mPMT (close to 1.5% coverage)
  - 11932 OD (OD photocoverage parameter value unchanged at 0.28)

We actually observe
- ~1x-~10x more killed tracks than the original WCSim/WCSim. The difference is more pronounced with the particle guns starting in the OD.
- Inconsistency in 20" & mPMT hits between the merged code & the nuPRISM/develop code for the hybrid geometry.

# Geometry test 2
And for the following geometry (closer to what we currently have as WCSim default)
  - 19826 20" (close to 20% coverage)
  - 9927 mPMT (close to 10% coverage)
  - 13220 OD (OD photocoverage parameter value unchanged at 0.28)

This gives
- 0 or 1 killed tracks for all my standard particle guns. This is success!
  - Recall that in the geometry with the extra tower, there are a three that give 10k's (up to 220k)
- Good consistency in terms of the standard validation plots, with the benefit of a reduced assymetry between the particle guns in the x/y directions

# So what now...
Obviously using the geometry with fewer mPMTs isn't a good idea, since the results are odd. Lets see if we can understand why
 
The following links have plots showing, in the usual rolled-out cylinder 2D format (note that the left half displays both caps - they are laid out identically by construction in WCSim)
- page 1 = 20" PMT layout
- page 2 = OD PMT layout
- page 3 = mPMT layout

[1550 mPMT geometry / test 1](https://github.com/spradlin/WCSim/files/11633809/display_geo_noextratower.pdf)
- mPMTs are laid out in stripes, that will not be how mPMTs are laid out in the real detector. The stripes are such that there are areas of "high" and "low" density mPMTs, which makes some of the odd results seen more understandable
  - It doesn't however explain the large number of killed tracks
- There are corresponding stripes in the 20" layout, where the 20" PMTs are missing. Again, it is my understanding that this is not how the real detector will be laid out

[9927 mPMTs geometry / test 2](https://github.com/spradlin/WCSim/files/11633811/display_geo_noextratower2.pdf)
- The striped features described above all still there, but things are less pronounced

Therefore the recommendation is to use the 9927 mPMT geometry as the new WCSim default, and implement hit masking.  
People can of course take a look at tweaking WCSim parameters to produce hybrid ID geometries that have fewer mPMTs, with PMTs laid out in a sensible pattern. If you are doing this, I strongly recommend 


# TLDR;
- I've set the default number of PMTs to
  - 19826 20"
  - 9927 mPMT
  - 13220 OD (OD photocoverage parameter value unchanged at 0.28)
- Building an extra tower now comes with just a warning (it isn't fatal)
- This is all in the new tag!
